kernel/constr.mli

```diff
@@ -238,6 +241,7 @@ type ('constr, 'types, 'sort, 'univs) kind_of_term =
   | Proj      of Projection.t * 'constr
   | Int       of Uint63.t
   | Float     of Float64.t
+  | Array     of 'univs * 'constr array * 'constr * 'types
```

Comment:
Something that I've been wondering for a while is if `constr` should not group all primitive types under a `Prim` constructor; this seems a bit more scalable, as of today adding a new primitive type is quite invasive I think.

Reply:
>Something that I've been wondering for a while is if constr should not group all primitive types under a Prim constructor; this seems a bit more scalable, as of today adding a new primitive type is quite invasive I think.

I don't see how it would help. Since arrays contain constrs all the folds would still need updating for instance

Reply:
You would need to update `fold_prim` indeed; I was thinking more on the lines of "imagine I want to add today 'primitive integers'"

Can it be done in a non-invasive way?

Reply:
Why make strings primitive, rather than define them using arrays and integers?

Reply:
Indeed, I mostly meant them as an example. It could be well the case that no more primitive stuff is needed so it'd be pointless to wonder about this.

Reply:
I think the question is not 'would it be useful to have anything else?' but 'how much should we include in the Coq kernel, and how and why?'  After chatting with @andres-erbsen, it seems there are other data structures were imperative implementations can give useful speedups.  For example, it seems union-find requires interesting imperative features (cf [Formalization of a persistent union-find data structure in Coq](https://www.lri.fr/~filliatr/puf/)), and I'm told that representing bignums via arrays or linked lists of ints from within Coq would break the CPU's ability to perform out-of-order execution (due to double-indirection of pointers).  (Unclear if we care about this level of performance.)  I wonder if @ppedrot has thoughts on this.

Reply:
I'm not saying there are no more interesting candidates for primitive data structures (I don't know). Anyway, I don't see how adding a `Primitive` constructor would help. Each iterator, comparison function, reduction machine, typing machine, would still have to be adapted.